### PR TITLE
optional task 01

### PR DESCRIPTION
**Pull request by PavelYudzin**

### CODE REVIEW

File: java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+        for (int val : intArray) {
+            int[] digitsOfTheNumber;
+            boolean hasRisingDigitsSequence = false;
+            digitsOfTheNumber = getDigitsOfTheNumber(val);
+
+            if (String.valueOf(val).length() == 1) {
+                continue;
+            }
+
+            for (int i = 0; i < digitsOfTheNumber.length - 1; i++) {
```
Comment (VitaliLebedzeu) [inline java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java]: Try to avoid inner loops

Comment (PavelYudzin) [reply java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java]: All loops were corrected. Inner loops were executed as additional methods.

File: java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+import java.util.Scanner;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class OptionalTask01 {
+    public static void main(String[] args) {
+        String numbersLine;
+        String[] stringOfNumbers;
+
+        numbersLine = inputIntegersAsAString();
```
Comment (VitaliLebedzeu) [inline java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java]: Please mix the declaration and initialization in one line.

File: java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+
+        stringOfNumbers = numbersLine.split("\\s+");
```
Comment (VitaliLebedzeu) [inline java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java]: Please mix the declaration and initialization in one line.

File: java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+
+
+    private static String inputIntegersAsAString() {
+        boolean hasWrongCharacters;
+        boolean hasDigits;
+        String line;
+        String message;
+        Scanner in = new Scanner(System.in);
+
+        message = "Enter integer numbers using SPACE to divide them:>> ";
```
Comment (VitaliLebedzeu) [inline java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java]: Please mix the declaration and initialization in one line.

File: java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+        System.out.println("The first number with rising digits sequence: " + integerWithRisingDigitsSequence + "\n");
+
+        int integerWithOnlyDifferentDigits;
+        integerWithOnlyDifferentDigits = getIntegerWithOnlyDifferentDigits(stringOfNumbers);
+        System.out.println("The first number with only different digits: " + integerWithOnlyDifferentDigits + "\n");
+
+    }
+
+
+    private static String inputIntegersAsAString() {
```
Comment (VitaliLebedzeu) [inline java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java]: This method should start as get...

File: java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java [hunk 6]
```diff
@@ -0,0 +1,414 @@
...
+        Scanner in = new Scanner(System.in);
+
+        message = "Enter integer numbers using SPACE to divide them:>> ";
+        do {
+            System.out.print(message);
+            Pattern restrictedCharacters = Pattern.compile("[^0-9-\\s+]");
+            Pattern mustHave = Pattern.compile("[\\d]");
+            line = in.nextLine();
+            Matcher matcher = restrictedCharacters.matcher(line);
+            hasWrongCharacters = matcher.find();
```
Comment (VitaliLebedzeu) [inline java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java]: Write this like 
hasWrongCharacters = restrictedCharacters.matcher(line).find();
or
hasWrongCharacters = Pattern.compile("[^0-9-\\s+]").matcher(line).find();

File: java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java [hunk 7]
```diff
@@ -0,0 +1,414 @@
...
+            matcher = mustHave.matcher(line);
+            hasDigits = matcher.find();
```
Comment (VitaliLebedzeu) [inline java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java]: the same as https://github.com/PavelYudzin/test-automation-training/pull/2/files#r525371515

File: java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java [hunk 8]
```diff
@@ -0,0 +1,414 @@
...
+        } while (hasWrongCharacters || !hasDigits);
+
+        line = line.replaceAll("\\+\\s*", " ");
+        line = line.replaceAll("-\\s*", " -");
+        line = line.trim();
+        return line;
```
Comment (VitaliLebedzeu) [inline java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java:83]: you can write this like 
return line.trim();

File: java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java [hunk 9]
```diff
@@ -0,0 +1,414 @@
...
+    }
+
+
+    private static void printTheShortestNumber(String[] numbers) {
+        int minLength;
+        minLength = numbers[0].length();
```
Comment (VitaliLebedzeu) [inline java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java]: Please mix the declaration and initialization in one line.

File: java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java [hunk 10]
```diff
@@ -0,0 +1,414 @@
...
+
+        for (String item : numbers) {
+            if (item.length() < minLength) {
+                minLength = item.length();
+            }
+        }
+        System.out.print("The shortest number(s) is(are): ");
```
Comment (VitaliLebedzeu) [inline java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java:95]: Recommend: Separate loop into another method like 
private int getMinNumberLength(String[] numbers) {
    int minLength = numbers[0].length;
    for (String item : numbers) {
         if (item.length() < minLength) {
        minLength = item.length();
        }
    }
    System.out.print("The shortest number(s) is(are): ");
return minLength;
}

File: java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java [hunk 11]
```diff
@@ -0,0 +1,414 @@
...
+
+        for (String val : numbers) {
+            if (val.length() == minLength) {
+                System.out.print(val + " ");
+            }
+        }
+        System.out.print("(length: " + minLength + ")\n");
+    }
```
Comment (VitaliLebedzeu) [inline java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java:103]: The same as https://github.com/PavelYudzin/test-automation-training/pull/2/files#r525384197

File: java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java (hunk 1 above)
Comment (PavelYudzin) [reply java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java]: All loops were fixed. Inner loops were executed as additional methods.

File: java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java [hunk 12]
```diff
@@ -0,0 +1,414 @@
...
+
+    private static boolean hasOnlyDifferentDigits(String number) {
+        boolean hasOnlyDifferentDigits = true;
+        StringBuilder sequence = new StringBuilder();
+
+        for (int i = 0; i < number.length(); i++) {
+            if (number.length() == 1) {
+                hasOnlyDifferentDigits = false;
+                break;
+            }
```
Comment (VitaliLebedzeu) [inline java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java]: Please move this if() from the for()

File: java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java [hunk 13]
```diff
@@ -0,0 +1,414 @@
...
+    }
+
+
+    private static boolean hasRisingDigitsSequence(String number) {
+        boolean isRisingDigitsSequence = false;
+        int[] digitsOfTheNumber = getDigitsOfTheNumber(number);
+
+        for (int i = 0; i < digitsOfTheNumber.length - 1; i++) {
+            if (digitsOfTheNumber[i + 1] > digitsOfTheNumber[i]) {
+                isRisingDigitsSequence = true;
```
Comment (VitaliLebedzeu) [inline java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java:346]: return true;

File: java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java [hunk 14]
```diff
@@ -0,0 +1,414 @@
...
+            } else {
+                isRisingDigitsSequence = false;
+                break;
```
Comment (VitaliLebedzeu) [inline java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java:349]: return false;

File: java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java [hunk 15]
```diff
@@ -0,0 +1,414 @@
...
+            }
+        }
+
+        return isRisingDigitsSequence;
```
Comment (VitaliLebedzeu) [inline java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java:353]: private static boolean hasRisingDigitsSequence(String number) {
        int[] digitsOfTheNumber = getDigitsOfTheNumber(number);

        for (int i = 0; i < digitsOfTheNumber.length - 1; i++) {
            if (digitsOfTheNumber[i + 1] < digitsOfTheNumber[i]) {
                return false;
            } 
        }
        return true;
}

File: java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java [hunk 16]
```diff
@@ -0,0 +1,414 @@
...
+    }
```
Comment (VitaliLebedzeu) [inline java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java:322]: `private static boolean hasRisingDigitsSequence(String number) {
        int[] digitsOfTheNumber = getDigitsOfTheNumber(number);

        for (int i = 0; i < digitsOfTheNumber.length - 1; i++) {
            if (digitsOfTheNumber[i + 1] < digitsOfTheNumber[i]) {
                return false;
            } 
        }
        return true;
}`

File: java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java (hunk 2 above)
Comment (PavelYudzin) [reply java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java]: Fixed

File: java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java (hunk 3 above)
Comment (PavelYudzin) [reply java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java]: Fixed

File: java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java (hunk 4 above)
Comment (PavelYudzin) [reply java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java]: Fixed

File: java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java (hunk 5 above)
Comment (PavelYudzin) [reply java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java]: Fixed

File: java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java (hunk 8 above)
Comment (PavelYudzin) [reply java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java:83]: I won't. It's not a big deal.
I suppose, If I do some manipulations with this String I should better  get this done before return.
I think it's more readable

File: java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java (hunk 9 above)
Comment (PavelYudzin) [reply java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java]: Fixed

File: java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java (hunk 6 above)
Comment (PavelYudzin) [reply java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java]: Fixed

File: java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java (hunk 7 above)
Comment (PavelYudzin) [reply java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java]: Fixed

File: java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java (hunk 12 above)
Comment (PavelYudzin) [reply java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java]: Fixed

File: java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java (hunk 16 above)
Comment (PavelYudzin) [reply java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java:322]: Fixed

File: java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java (hunk 10 above)
Comment (PavelYudzin) [reply java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java:95]: Rejected.

File: java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java (hunk 11 above)
Comment (PavelYudzin) [reply java/src/main/java/com/company/automation/fundamentals/optional_task/optional_task_01/OptionalTask01.java:103]: Rejected

**Review by VitaliLebedzeu — APPROVED** [review]

**Pull request merged**